File: langgraph-typescript/src/phaseLog.ts
import { AsyncLocalStorage } from "node:async_hooks";

import { Sandbox } from "./tools.js";

export const PHASES = [
  "researcher",
  "planner",
  "designer",
  "diagrammer",
  "illustrator",
  "coder",
  "reviewer",
] as const;

export const DEFAULT_PHASES = ["planner", "designer", "coder", "reviewer"] as const;

export type Phase = (typeof PHASES)[number];

const PREREQUISITES: Record<Phase, string[]> = {
  researcher: [],
  planner: [],
  designer: ["plan.md"],
  diagrammer: ["plan.md", "design.md"],
  illustrator: [],
  coder: ["plan.md", "design.md"],
  reviewer: ["plan.md", "design.md"],
};

const ARTIFACT_PRODUCERS: Record<string, Phase> = {
  "plan.md": "planner",
  "design.md": "designer",
  "research.md": "researcher",
};

export function parseAgents(value?: string | string[]): Phase[] {
  if (value == null) return [...DEFAULT_PHASES];
  const raw =
    typeof value === "string"
      ? value
          .replace(/,/g, " ")
          .split(/\s+/)
          .map((a) => a.trim().toLowerCase())
          .filter(Boolean)
      : value.map((a) => a.trim().toLowerCase()).filter(Boolean);
  if (!raw.length) return [...DEFAULT_PHASES];
  const unknown = raw.filter((a) => !PHASES.includes(a as Phase));
  if (unknown.length) {
    throw new Error(
      `Unknown agent(s): ${unknown.join(", ")}. Valid: ${PHASES.join(", ")}`,
    );
  }
  return PHASES.filter((p) => raw.includes(p));
}

function readArtifact(sandbox: Sandbox, filePath: string): string {
  const text = sandbox.readFile(filePath);
  return text.startsWith("ERROR") ? "" : text;
}

export function loadArtifacts(sandbox: Sandbox) {
  const written = sandbox.listWrittenFiles();
  return {
    plan: readArtifact(sandbox, "plan.md"),
    design: readArtifact(sandbox, "design.md"),
    review: readArtifact(sandbox, "review.md"),
    research: readArtifact(sandbox, "research.md"),
    files: written.filter((f) => f.startsWith("src/")),
    diagrams: written.filter((f) => f.startsWith("diagrams/") && f.endsWith(".mmd")),
    assets: written.filter((f) => f.startsWith("assets/") && f.endsWith(".png")),
  };
}

export function ensurePrerequisites(phase: Phase, sandbox: Sandbox): void {
  const missing: string[] = [];
  for (const artifact of PREREQUISITES[phase]) {
    if (sandbox.readFile(artifact).startsWith("ERROR")) missing.push(artifact);
  }
  if (phase === "reviewer") {
    const srcFiles = sandbox.listWrittenFiles().filter((f) => f.startsWith("src/"));
    if (!srcFiles.length) missing.push("src/**");
  }
  if (missing.length) {
    throw new Error(
      `Cannot run '${phase}': missing prerequisite(s): ${missing.join(", ")}. ` +
        "Run earlier phase(s) first or reuse --run-id with existing artifacts.",
    );
  }
}

export function validateAgentSelection(selected: Phase[], sandbox: Sandbox): void {
  const errors: string[] = [];
  for (const phase of selected) {
    for (const artifact of PREREQUISITES[phase]) {
      if (!sandbox.readFile(artifact).startsWith("ERROR")) continue;
      const producer = ARTIFACT_PRODUCERS[artifact];
      if (producer && selected.includes(producer) && selected.indexOf(producer) < selected.indexOf(phase)) {
        continue;
      }
      if (producer) {
        errors.push(
          `'${phase}' needs ${artifact} — add '${producer}' before it in --agents, ` +
            "or pass --run-id with existing artifacts",
        );
      } else {
        errors.push(`'${phase}' needs ${artifact}`);
      }
    }
    if (phase === "reviewer") {
      const srcFiles = sandbox.listWrittenFiles().filter((f) => f.startsWith("src/"));
      if (
        !srcFiles.length &&
        !(selected.includes("coder") && selected.indexOf("coder") < selected.indexOf("reviewer"))
      ) {
        errors.push(
          "'reviewer' needs src/** — add 'coder' before it in --agents, " +
            "or pass --run-id with existing code",
        );
      }
    }
  }
  if (errors.length) {
    throw new Error(`Invalid agent selection:\n- ${errors.join("\n- ")}`);
  }
}

export type PhaseEvent = {
  phase: string;
  index: number;
  total: number;
  run_id: string;
  framework: string;
};

type PhaseListener = (event: PhaseEvent) => void;

const phaseStore = new AsyncLocalStorage<PhaseListener>();

/** Run `fn` with a phase listener visible to `logPhaseStart` (HTTP SSE). */
export function runWithPhaseListener<T>(
  listener: PhaseListener,
  fn: () => Promise<T>,
): Promise<T> {
  return phaseStore.run(listener, fn);
}

export function logPhaseStart(
  phase: string,
  index: number,
  total: number,
  runId: string,
  framework: string,
): void {
  const banner =
    `\n${"═".repeat(40)}\n` +
    `Pipeline ${index}/${total} · ${phase}\n` +
    `(run_id=${runId} · ${framework})\n` +
    `${"═".repeat(40)}\n`;
  process.stderr.write(banner);
  phaseStore.getStore()?.({
    phase,
    index,
    total,
    run_id: runId,
    framework,
  });
}

File: crewai-python/app/phase_log.py
"""Pipeline phase logging and agent selection helpers."""

from __future__ import annotations

import sys
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from app.tools import Sandbox

PHASES = (
    "researcher",
    "planner",
    "designer",
    "diagrammer",
    "illustrator",
    "coder",
    "reviewer",
)

DEFAULT_PHASES = ("planner", "designer", "coder", "reviewer")

PREREQUISITES: dict[str, tuple[str, ...]] = {
    "researcher": (),
    "planner": (),
    "designer": ("plan.md",),
    "diagrammer": ("plan.md", "design.md"),
    "illustrator": (),
    "coder": ("plan.md", "design.md"),
    "reviewer": ("plan.md", "design.md"),
}

ARTIFACT_PRODUCERS: dict[str, str] = {
    "plan.md": "planner",
    "design.md": "designer",
    "research.md": "researcher",
}


def parse_agents(value: str | list[str] | None) -> list[str]:
    """Parse --agents flag or programmatic list into canonical order."""
    if value is None:
        return list(DEFAULT_PHASES)
    if isinstance(value, str):
        raw = [a.strip().lower() for a in value.replace(",", " ").split() if a.strip()]
    else:
        raw = [a.strip().lower() for a in value if a.strip()]
    if not raw:
        return list(DEFAULT_PHASES)
    unknown = set(raw) - set(PHASES)
    if unknown:
        raise ValueError(
            f"Unknown agent(s): {', '.join(sorted(unknown))}. "
            f"Valid: {', '.join(PHASES)}"
        )
    return [p for p in PHASES if p in raw]


def _read_artifact(sandbox: Sandbox, path: str) -> str:
    text = sandbox.read_file(path)
    if text.startswith("ERROR"):
        return ""
    return text


def load_artifacts(sandbox: Sandbox) -> dict[str, str | list[str]]:
    """Load existing artifacts from the sandbox."""
    plan = _read_artifact(sandbox, "plan.md")
    design = _read_artifact(sandbox, "design.md")
    review = _read_artifact(sandbox, "review.md")
    research = _read_artifact(sandbox, "research.md")
    files = [f for f in sandbox.list_written_files() if f.startswith("src/")]
    diagrams = [
        f for f in sandbox.list_written_files() if f.startswith("diagrams/") and f.endswith(".mmd")
    ]
    assets = [
        f for f in sandbox.list_written_files() if f.startswith("assets/") and f.endswith(".png")
    ]
    return {
        "plan": plan,
        "design": design,
        "review": review,
        "research": research,
        "files": files,
        "diagrams": diagrams,
        "assets": assets,
    }


def ensure_prerequisites(phase: str, sandbox: Sandbox) -> None:
    """Raise ValueError if required artifacts are missing for a phase."""
    missing: list[str] = []
    for artifact in PREREQUISITES.get(phase, ()):
        text = sandbox.read_file(artifact)
        if text.startswith("ERROR"):
            missing.append(artifact)
    if phase == "reviewer":
        src_files = [f for f in sandbox.list_written_files() if f.startswith("src/")]
        if not src_files:
            missing.append("src/**")
    if missing:
        raise ValueError(
            f"Cannot run '{phase}': missing prerequisite(s): {', '.join(missing)}. "
            f"Run earlier phase(s) first or reuse --run-id with existing artifacts."
        )


def validate_agent_selection(selected: list[str], sandbox: Sandbox) -> None:
    """Fail fast when selected agents cannot satisfy prerequisites."""
    errors: list[str] = []
    for phase in selected:
        for artifact in PREREQUISITES.get(phase, ()):
            if not sandbox.read_file(artifact).startswith("ERROR"):
                continue
            producer = ARTIFACT_PRODUCERS.get(artifact)
            if producer and producer in selected and selected.index(producer) < selected.index(phase):
                continue
            if producer:
                errors.append(
                    f"'{phase}' needs {artifact} — add '{producer}' before it in --agents, "
                    f"or pass --run-id with existing artifacts"
                )
            else:
                errors.append(f"'{phase}' needs {artifact}")
        if phase == "reviewer":
            src_files = [f for f in sandbox.list_written_files() if f.startswith("src/")]
            if not src_files and not (
                "coder" in selected and selected.index("coder") < selected.index("reviewer")
            ):
                errors.append(
                    "'reviewer' needs src/** — add 'coder' before it in --agents, "
                    "or pass --run-id with existing code"
                )
    if errors:
        raise ValueError("Invalid agent selection:\n- " + "\n- ".join(errors))


def log_phase_start(
    phase: str,
    index: int,
    total: int,
    run_id: str,
    framework: str,
) -> None:
    """Print a pipeline-level banner to stderr before each agent phase."""
    banner = (
        f"\n{'═' * 40}\n"
        f"Pipeline {index}/{total} · {phase}\n"
        f"(run_id={run_id} · {framework})\n"
        f"{'═' * 40}\n"
    )
    sys.stderr.write(banner)
    sys.stderr.flush()
    try:
        from feature_delivery_api.progress import emit_phase
    except ImportError:
        return
    emit_phase(phase, index, total, run_id, framework)
